Re-enable preprocessing and fix template overloads

Maetveis opened the pull request:
HI, sorry for the unusual workflow, I tried your PR (https://github.com/ROCmSoftwarePlatform/rocPRIM/pull/336) on rocPRIM and I made some changes to improve it. As it is a larger changeset I think this way its easier than adding comments and suggestions. (I will do that for the smaller parts).

Commit:
Re-enable preprocessing for doxygen

`__host__` and `__device__` function attributes are shown in the docs.

Re-enable preprocessing for doxygen, adds cuda / hip attributes
to `cpp_id_attributes` in sphinx this fixes it erroring out on these
attributes.
The rocprim namespace is added for all function name references in the
rst docs, they were not working previously because the namespace macros
were not expanded by doxygen.

Commit:
Fix documentation of template overloaded methods with doxygen groups

Use doxygen groups as a workaround for breathe failing to find methods
that have complex template names, or both templated and non templated
overloads.

v01dXYZ commented:
I rebased your modifications. I reformulated a little bit your commit messages to help me understand the intent of the modifications. Thank you for the configuration of Doxygen for macro expansion and also the `cpp_id_attributes` which I don't even know it existed.

Tell me if you're happy with it and I'll close the MR if it is the case.

Maetveis commented:
> I rebased your modifications. I reformulated a little bit your commit messages to help me understand the intent of the modifications. Thank you for the configuration of Doxygen for macro expansion and also the `cpp_id_attributes` which I don't even know it existed.
> 
> Tell me if you're happy with it and I'll close the MR if it is the case.

I added one comment there, but yes it is okay, let me close this.
